### PR TITLE
Cache address-to-ActorType translation

### PR DESCRIPTION
**Pull request by mb1896**
This closes ENG-616

### CODE REVIEW

**Comment by mb1896** [comment]
@aakoshh Please take a look when you get a chance!

File: fendermint/eth/api/src/cache.rs [hunk 1]
```diff
@@ -107,4 +110,54 @@ where
...
+    use crate::state::ActorType;
+    use cid::Cid;
+    use fendermint_rpc::FendermintClient;
+    use fvm_shared::address::Address;
+    use std::str::FromStr;
+    use tendermint_rpc::HttpClient;
+
+    #[test]
+    fn test_read_and_write_addr_to_actor_type() {
+        let client = FendermintClient::new(HttpClient::new("http://localhost").unwrap());
```
Comment (aakoshh) [inline fendermint/eth/api/src/cache.rs:137]: There is a `MockClient` for them, e.g. [here](https://github.com/consensus-shipyard/ipc/blob/c3e4c020d52decfbc57eb6e2618c947f3f7b29a7/fendermint/vm/snapshot/src/manager.rs#L483).

File: fendermint/eth/api/src/state.rs [hunk 2]
```diff
@@ -442,6 +443,12 @@ where
         height: FvmQueryHeight,
     ) -> JsonRpcResult<ActorType> {
         let addr = to_fvm_address(*address);
+
+        if let Some(actor_type) = self.addr_cache.get_actor_type_from_addr(&addr) {
+            log::debug!("cache contains the result: {:?}, return it", actor_type);
```
Comment (aakoshh) [inline fendermint/eth/api/src/state.rs:448]:
```suggestion
            tracing::debug!(addr, actor_type, "cache hit");
```
I'm not sure if there's a difference between using `tracing::log::debug!` but usually we use `tracing::debug!`. The macro supports structured logging like `debug!(primitive_value, key2 %= value_with_display, ?value_with_debug, "static message")` which means those values will turn up in a way that is easier to filter for in log aggregators. We should prefer them to interpolating values into the message.

File: fendermint/eth/api/src/state.rs [hunk 3]
```diff
@@ -457,6 +464,8 @@ where
             Some((typ, _)) => ActorType::Known(Cow::Owned(typ)),
             None => ActorType::Unknown(actor_type_cid),
         };
+        log::debug!("put result into cache: {:?} => {:?}", addr, ret);
```
Comment (aakoshh) [inline fendermint/eth/api/src/state.rs:467]: Here too you can use structured logging.

**Review by aakoshh — COMMENTED** [review]
Nice! Have you considered caching the CID-to-type relationship as well? 

It makes sense to store something on the address level because that's what we are asking about, and first we have to retrieve the state of the actor to get its code CID, however, there are only a handful of CIDs, so caching that second lookup could eliminate half of the queries on its own.

**Pull request merged**